crates/env/src/backend.rs

```diff
@@ -183,6 +186,16 @@ pub trait EnvBackend {
     /// Clears the contract's storage key entry.
     fn clear_contract_storage(&mut self, key: &Key);
 
+    fn set_storage_value<V>(&mut self, key: &StorageKey, value: &V) -> Option<u32>
+    where
+        V: scale::Encode;
+
+    fn get_storage_value<R>(&mut self, key: &StorageKey) -> Result<Option<R>>
+    where
+        R: scale::Decode;
+
+    fn clear_storage_value(&mut self, key: &StorageKey);
```

Comment:
I forgot to add comments=(I will fix that.

`Mapping` hashes the `Key` with the storage key and gets 32 bytes old storage key. We can easily convert it to a new storage key with 4 bytes and remove duplicate methods. The problem is that it will increase the chance of collision(we reduced 32 bytes to 4 bytes).

With transparent hashing of course we will have only one bunch of methods. 

I will prepare the code for transparent hashing and remove duplicates.

The signature will be next:
```rust
fn set_contract_storage<V>(&mut self, storage_key: &StorageKey, key: Option<&Key>, value: &V) -> Option<u32>
where
    K: scale::Encode;
    V: scale::Encode;
```

If the `key` is none, we store the `value` under `storage_key`. Otherwise, hash `key` and  `storage_key` like in `Mapping`. With transparent hashing we will remove hashing of keys)

Reply:
Done. Now we don't have `Key` at all.

I moved the hashing logic into `set_contract_storage`, `get_contract_storage` and etc.
But it increased the size of the contracts(`Erc20`, `Erc721`, `Erc1155`) again by ~100 bytes.

But after some investigation, I found that inlining can help again=) Now the size is reduced by ~230 bytes=D

```rust
#[inline(always)]
/// Returns storage key.
fn storage_key<K>(
    &mut self,
    storage_key: &StorageKey,
    dynamic_key: Option<K>,
) -> [u8; 32]
where
    K: scale::Encode,
{
    if let Some(dynamic_key) = dynamic_key {
        self.combined_storage_key(storage_key, dynamic_key)
    } else {
        StorageKeyComposer::old_key(storage_key)
    }
}

#[inline(never)]
/// Combines the storage key with dynamic key knows at runtime.
fn combined_storage_key<K>(
    &mut self,
    storage_key: &StorageKey,
    dynamic_key: K,
) -> [u8; 32]
where
    K: scale::Encode,
{
    let mut hash = <Blake2x256 as HashOutput>::Type::default();
    self.hash_encoded::<Blake2x256, _>(&(dynamic_key, storage_key), &mut hash);
    hash
}
```

Reply:
Nice. I look it much better know. The details of the actual key passed to pallet-contracts are now contained to the lower levels.

I think the issues about code size will go away once we stop messing around with `[u8; 32]` which trigger a lot of copy code to pass around. Once transparent hashing is implemented we don't hash but just concat that stuff into a `Vec<u8>` which can be moved by pointer.